### ISSUE TITLE
Parallel tests are hunging with 100% CPU load

### ISSUE THREAD

**Issue by OsirisTerje** [nunit/nunit#2507]
@lamest commented on [Tue Oct 17 2017](https://github.com/nunit/nunit3-vs-adapter/issues/409)

If too many async tests runs in parallel, part on tests (usually 4 or less) runs normally and the rest of them pending indefinitely with CPU load near 100%

* NUnit 3.8.1 and NUnit3TestAdapter 3.8.0
* Visual Studio Enterprise 15.3.5 
* [repro](https://github.com/lamest/NunitParallelTest.git)
* .net Framework 4.6.1

---

@OsirisTerje commented on [Tue Oct 17 2017](https://github.com/nunit/nunit3-vs-adapter/issues/409#issuecomment-337211364)

Can you please upload a repro project?
Make it as small as you can, it should compile "as is".

---

@lamest commented on [Wed Oct 18 2017](https://github.com/nunit/nunit3-vs-adapter/issues/409#issuecomment-337453744)

@OsirisTerje link was in description. You need zip?

---

@rogierpoldner commented on [Wed Oct 18 2017](https://github.com/nunit/nunit3-vs-adapter/issues/409#issuecomment-337560631)

I have the same issue, it seems as if 3.7.1 was the latest version working correctly, version 3.8.0 and 3.8.1. result in 'hung' of the tests. Also running without parallelism works correctly in 3.8.0 and on wards. The test run with parallelism in 3.8.0 (and 3.8.1) stops exactly after 6 tests with a parallelism of 3 tests, after that no new tests are started anymore. 

I am using NUnit in combination with Selenium for automated testing. I have the following configured in AssemblyInfo to run in parallel:

```
[assembly: Parallelizable(ParallelScope.Fixtures)]
[assembly: LevelOfParallelism(3)]
```

---

@OsirisTerje commented on [Wed Oct 18 2017](https://github.com/nunit/nunit3-vs-adapter/issues/409#issuecomment-337575478)

@lamest Ahh, sorry, didn't see that link.  Thanks!! (and git repo as you did is perfect and preferred)
And confirmed......

---

@OsirisTerje commented on [Wed Oct 18 2017](https://github.com/nunit/nunit3-vs-adapter/issues/409#issuecomment-337585458)

@rprouse @CharliePoole This seems to be a NUnit engine or framework error.  The run in VS can be cancelled, and it then also stops the executionengine.  
Same thing happens with NUnit Console, and I checked both with 3.7 and 3.5 versions, and also with Resharper runner.  The number of tests it runs differs between VS and R#, and resharper runner doesn't die when run is cancelled.    

Should this be moved to NUnit  repo?

Some screenshots:

![image](https://user-images.githubusercontent.com/203432/31720300-769ce670-b416-11e7-8a5c-97b7333bf54c.png)

![image](https://user-images.githubusercontent.com/203432/31720309-7db7bc5a-b416-11e7-815f-43e445618c7b.png)

![image](https://user-images.githubusercontent.com/203432/31720318-84b26f64-b416-11e7-82ff-0ee7d6df5c8a.png)

![image](https://user-images.githubusercontent.com/203432/31720321-89b1a2fa-b416-11e7-8322-c37ab63d8d24.png)

![image](https://user-images.githubusercontent.com/203432/31720332-94dba2c0-b416-11e7-80ce-f50b6db2ed20.png)

![image](https://user-images.githubusercontent.com/203432/31720341-9adcd7de-b416-11e7-8e59-9d37b0306cf5.png)

![image](https://user-images.githubusercontent.com/203432/31720347-9f74bc12-b416-11e7-8ee3-1a5e3c84d4f7.png)

---

@CharliePoole commented on [Wed Oct 18 2017](https://github.com/nunit/nunit3-vs-adapter/issues/409#issuecomment-337664526)

@OsirisTerje Yes, this is part of the general set of problems we are encountering with parallelism in the framework. Since some of them are relatively new, users may gain some relief by going back to an older version of the framework. This should be moved to the nunit repo, where we can determine if it represents a new aspect of the problem or duplicates existing issues.

[I'm not using ZenHub at the moment or I would move it myself.]

**Comment by liamharries** [comment]
Facing the same issue with versions 3.8.1 & 3.8.0, reverting to 3.7.1 or previous removes the issue.

**Comment by ChrisMaddock** [comment]
@CharliePoole's been working on some of these bugs in https://github.com/nunit/nunit/pull/2476. 

@CharliePoole - can I assign this issue to you, along with the rest? This seems a different case in that it uses lots of async, but I assume in the same area?

**Comment by CharliePoole** [comment]
@OsirisTerje I'd like to wait and see if PR #2476 fixes this.

**Comment by lamest** [comment]
looks like the repro project works fine when updated to NUnit 3.9.0 and NUnit3TestAdapter 3.9.0. Thanks!)

**Comment by rogierpoldner** [comment]
My issues were also solved with the new version. Thanks for the quick support!

**Comment by rprouse** [comment]
Closing as fixed in the 3.9 release
